### PR TITLE
client-go/testing: properly handle Patch actions for missing Objects

### PR DESCRIPTION
**Pull request by bouk**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Currently the fake client will return a default empty Object when a Patch action is submitted on a missing Object. The correct behavior is to instead propagate the NotFound error.

**Does this PR introduce a user-facing change?**:
```release-note:
client-go: fake clients now properly return NotFound errors when attempting to patch non-existent objects
```

### CODE REVIEW

**Comment by bouk** [comment]
It seems the test failure is an issue in some other part of the faking code, as it can’t find the ReplicaSets when doing the patch therefore the adopting fails.

File: pkg/controller/controller_utils.go
Comment (roycaihw) [inline pkg/controller/controller_utils.go]: I don't understand why this need to be changed. The [test failure](https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/70886/pull-kubernetes-bazel-test/67258/) might just be a flake. Could you discard the change and retest?

File: staging/src/k8s.io/client-go/testing/fixture.go
Comment (roycaihw) [inline staging/src/k8s.io/client-go/testing/fixture.go]: LGTM. The previous code seemed to be an oversight. 

@caesarxuchao Does the boolean exist only to allow us skipping some valid actions in [chain](https://github.com/kubernetes/kubernetes/blob/3c5c6025bf5e962dd67931fd38e18031108d5a37/staging/src/k8s.io/client-go/testing/fake.go#L141-L143) which don't have reaction defined (Watch, ProxyGet)?

Comment (bouk) [reply staging/src/k8s.io/client-go/testing/fixture.go]: If you look at the rest of the function you’ll see that it’s only used for unknown actions, so yes it seems to be an oversight.

File: pkg/controller/controller_utils.go
Comment (bouk) [reply pkg/controller/controller_utils.go]: It is not a flake unfortunately, I was able to reproduce it locally. The reason the test fails is because it uses this helper, which uses the old ExtensionsV1Beta1 interface to RS, which *I think* causes the fake client to be unable to find the AppsV1 RS. The result is that when the deployment tries to patch the RS to adopt it, it can’t find it so it ignores it. 

The reason it worked before is because if the action wasn’t handled, it would return an empty object with a nil error, so it would ‘succeed’

File: staging/src/k8s.io/client-go/testing/fixture.go
Comment (yue9944882) [reply staging/src/k8s.io/client-go/testing/fixture.go]: surprise! 🙃unhandling patch events here makes the fake client working correctly w/ the cohibitating resources! the fake client is simply grouping requesting objects by its `gvr` so the test is actually suppose to be broken when we write w/ `extensions/v1` but read w/ `apps/v1`.

File: pkg/controller/controller_utils.go
Comment (yue9944882) [reply pkg/controller/controller_utils.go]: mmmmmm.. i guess it should have been done in #70370 @liggitt

**Review by yue9944882 — CHANGES_REQUESTED** [review]
please hold until @liggitt comments. thx!

/hold

File: pkg/controller/controller_utils.go
Comment (liggitt) [reply pkg/controller/controller_utils.go]: fake clients don't understand that apps and extensions replicasets are stored in the same place, nor would I expect them to. https://github.com/kubernetes/kubernetes/pull/70370 consistently moves all callers to apps/v1, let's defer this change to that PR

Comment (roycaihw) [reply pkg/controller/controller_utils.go]: Thanks for the explanation. IIUC the `fixture.go` change in this PR is legit, but we are bitten by our debt of having two API groups share the storage, and the code migration being half-way. 

@bouk We can rebase after #70370 is merged, then apply the labels to merge this PR.

Comment (bouk) [reply pkg/controller/controller_utils.go]: Sounds good!

File: staging/src/k8s.io/client-go/testing/fixture_test.go
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture_test.go]: "expected IsNotFound error" seems to be a better error message.

File: pkg/controller/controller_utils.go
Comment (caesarxuchao) [reply pkg/controller/controller_utils.go]: It won't cause conflict in #70370, as the [change](https://github.com/kubernetes/kubernetes/pull/70370/files#diff-44bba2c68a6ea2b4ca0dbb7f4465828e) there is identical, so I won't block this PR on #70370.

**Review by caesarxuchao — COMMENTED** [review]
lgtm except for a nit.

File: staging/src/k8s.io/client-go/testing/fixture_test.go
Comment (bouk) [reply staging/src/k8s.io/client-go/testing/fixture_test.go]: Switched it to use `EqualError` :+1:

**Comment by bouk** [comment]
I have amended to test to use a clearer assert, and rebased on master. In my opinion this is good to go.

File: pkg/controller/controller_utils.go
Comment (liggitt) [reply pkg/controller/controller_utils.go]: verified that all callers of this method are doing things that apply equally well to apps/v1 and extensions/v1beta1 replicasets

**Comment by liggitt** [comment]
lgtm

needs `hack/update-bazel.sh` run and results committed to make the build/tests pass

**Comment by bouk** [comment]
The new test fail seems legit, but it's because the fake api client now actually does patches. Not sure how to proceed

**Comment by caesarxuchao** [comment]
RE. new failing unit test, I asked a question in https://github.com/kubernetes/kubernetes/pull/70876/files#r243047547. The unit test seems wrong.

**Comment by liggitt** [comment]
correct, that test is invalid. it should be updated to patch a different field in the node object

**Comment by bouk** [comment]
I've rebased on the latest master and the test now succeeds

**Comment by bouk** [comment]
I rebased and the flaky tests now pass

**Pull request merged**
